File: objectivec-spritekit/Endless Runner/view/Far.m
//
//  Far.m
//  Endless Runner
//
//  Created by Christopher Caleb on 15/04/2014.
//  Copyright (c) 2014 Christopher Caleb. All rights reserved.
//

#import "Far.h"

static const CGFloat FAR_DELTA_X = 0.32;

@implementation Far
{
    SKTexture *texture;
    SKSpriteNode *far1;
    SKSpriteNode *far2;
    CGFloat viewportX;
}

- (id)init
{
    self = [super init];
    if (self) {
        texture = [SKTexture textureWithImageNamed:@"bg-far"];
        far1 = [SKSpriteNode spriteNodeWithTexture:texture];
        far2 = [SKSpriteNode spriteNodeWithTexture:texture];
        
        far1.anchorPoint = CGPointMake(0.0, 0.0);
        far1.position = CGPointMake(0, 256.0);
        [self addChild:far1];
        
        far2.anchorPoint = CGPointMake(0.0, 0.0);
        far2.position = CGPointMake(1024.0, 256.0);
        [self addChild:far2];
        
        viewportX = 0.0;
    }
    
    return self;
}

- (void)setViewportX:(CGFloat)newViewportX
{
    float distanceTravelled = newViewportX - viewportX;
    viewportX = newViewportX;
    
    CGPoint tempPoint = self.position;
    tempPoint.x -= (distanceTravelled * FAR_DELTA_X);
    if (tempPoint.x < -1024)
    {
        tempPoint.x += 1024;
    }
    
    self.position = tempPoint;
}

@end

File: swift-spritekit/Endless-Runner/Far.swift
//
//  Far.swift
//  Endless-Runner
//
//  Created by Christopher Caleb on 12/06/2014.
//  Copyright (c) 2014 Christopher Caleb. All rights reserved.
//

import Foundation
import SpriteKit

class Far: SKNode {
    let FAR_DELTA_X: CGFloat = 0.32
    
    let _texture: SKTexture
    let _far1: SKSpriteNode
    let _far2: SKSpriteNode
    var _viewportX: CGFloat
    
    init() {
        _texture = SKTexture(imageNamed: "bg-far")
        _far1 = SKSpriteNode(texture: _texture)
        _far2 = SKSpriteNode(texture: _texture)
        _viewportX = 0.0
        super.init()
        
        _far1.anchorPoint = CGPoint(x: 0.0, y: 0.0)
        _far1.position = CGPoint(x: 0.0, y: 256.0)
        addChild(_far1)
        
        _far2.anchorPoint = CGPoint(x: 0.0, y: 0.0)
        _far2.position = CGPoint(x: 1024.0, y: 256.0)
        addChild(_far2)
    }
    
    func setViewportX(newViewportX: CGFloat) {
        var distanceTravelled = newViewportX - _viewportX
        _viewportX = newViewportX
        
        var tempPoint = self.position;
        tempPoint.x -= distanceTravelled * FAR_DELTA_X
        if tempPoint.x < -1024 {
            tempPoint.x += 1024
        }
        position = tempPoint
    }
}
